test: enhance DummyExecutor

tzdybal opened the pull request:
<!--
Please read and fill out this form before submitting your PR.

Please make sure you have reviewed our contributors guide before submitting your
first PR.

NOTE: PR titles should follow semantic commits: https://www.conventionalcommits.org/en/v1.0.0/
-->

## Overview

Resolves #24

<!-- 
Please provide an explanation of the PR, including the appropriate context,
background, goal, and rationale. If there is an issue with this information,
please provide a tl;dr and link the issue. 

Ex: Closes #<issue number>
-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a method to inject transactions into the executor.
	- Updated transaction retrieval to return an empty list instead of nil.
	- Launched a gRPC server for the dummy executor, enabling remote execution capabilities.
	- Added robust error handling and signal management for server operations.

- **Bug Fixes**
	- Enhanced error handling for finalizing transactions with invalid block heights.

- **Tests**
	- Refined test assertions for transaction retrieval and finalization methods to align with new expected behaviors.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tzdybal committed:
test: enhance DummyExecutor

tzdybal committed:
chore: fix linter errors

MSevey commented on cmd/dummy/main.go:
```diff
@@ -0,0 +1,59 @@
+	pb "github.com/rollkit/go-execution/types/pb/execution"
+)
+
+const bufSize = 1024 * 1024
```
any reason behind these numbers we should capture?

tzdybal replied:
Sorry, mostly copy-paste ;)

MSevey commented on cmd/dummy/main.go:
```diff
@@ -0,0 +1,59 @@
+	go func() {
+		log.Println("Serving...")
+		if err := s.Serve(listener); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			log.Fatalf("Server exited with error: %v\n", err)
```
Doesn't `log.Fatal` kill the program at that line? Does this cause the program to hang because it doesn't then hit the `wg.Done()`?

MSevey reviewed (dismissed):
I think there are some clean up stuff like defining defaults for the listen address but find to approve and merge in to avoid blocking other work.

yarikbratashchuk commented on cmd/dummy/main.go:
```diff
@@ -0,0 +1,59 @@
+		MaxRequestSize: bufSize,
+	}
+
+	listenAddress := "127.0.0.1:40041"
+	if len(os.Args) == 2 {
+		listenAddress = os.Args[1]
+	}
```
Suggestion is good

yarikbratashchuk commented on cmd/dummy/main.go:
```diff
@@ -0,0 +1,59 @@
+		MaxRequestSize: bufSize,
+	}
+
+	listenAddress := "127.0.0.1:40041"
+	if len(os.Args) == 2 {
+		listenAddress = os.Args[1]
+	}
+
```
this one is good as well

yarikbratashchuk commented on cmd/dummy/main.go:
```diff
@@ -0,0 +1,59 @@
+	s := grpc.NewServer()
+	pb.RegisterExecutionServiceServer(s, server)
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	go func() {
+		log.Println("Serving...")
+		if err := s.Serve(listener); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			log.Fatalf("Server exited with error: %v\n", err)
+		}
+		wg.Done()
+	}()
+	defer s.Stop()
+
+	wg.Wait()
+	log.Println("Server stopped")
+}
```
would suggest the same

yarikbratashchuk reviewed (changes requested):
I think we should apply suggestions

tzdybal commented on cmd/dummy/main.go:
```diff
@@ -0,0 +1,59 @@
+		_ = listener.Close()
+	}()
+
+	log.Println("Starting server...")
+	server := grpcproxy.NewServer(dummy, config)
+	s := grpc.NewServer()
+	pb.RegisterExecutionServiceServer(s, server)
+
```
This is not intended for production use - there is no reason to setup TLS.

tzdybal committed:
feat: handle signals, use channel instead of WaitGroup

tzdybal committed:
refactor: improve flag parsing and logs

tzdybal committed:
fix: remove block from pending after finalization

This prevents from memory leaking and makes it impossible to `SetFinal`
more than once.
